libraries/chain/db_block.cpp

```diff
@@ -103,6 +140,32 @@ std::vector<block_id_type> database::get_block_ids_on_fork(block_id_type head_of
   result.emplace_back(branches.first.back()->previous_id());
   return result;
 }
+    
+void database::check_tansaction_for_duplicated_operations(const signed_transaction& trx)
+{
+   const auto& proposal_index = get_index<proposal_object>();
+   std::set<fc::sha256> existed_operations_digests;
+   
+   proposal_index.inspect_all_objects( [&](const object& obj){
```

Comment:
This is horribly inefficient and should be implemented as a secondary index instead. You could be calculating hundreds of hash values for every single `broadcast_transaction` call.

Reply:
True, we'll create an index that contains proposals indexed by account and only go through those that have witness-account affected. That way we reduce the range drastically.

Reply:
So I am trying to implement a new index for the `proposal_index`. AFAIU we want to iterate over proposals, where `WITNESS_ACCOUNT` is a member of `required_active_approvals`. 

@pmconrad 
Do you have any suggestions how this could be properly implemented?

Reply:
See `account_member_index` for an example how a secondary index works.
I'd add a secondary index to `proposal_index` that maintains a list of digests of operations that are currently contained in proposals.